File: 30-Day LeetCoding Challenge/Python/bstFromPreorder.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def bstFromPreorder(self, preorder):
        """
        :type preorder: List[int]
        :rtype: TreeNode
        """
        root = TreeNode(preorder[0])
        size = len(preorder)
        i = 1
        s = []
        s.append(root)
        while i < len(preorder):
            temp = None
            while(len(s) > 0 and s[-1].val < preorder[i]):
                temp = s.pop()
            if temp != None:
                temp.right = TreeNode(preorder[i])
                s.append(temp.right)
            else:
                temp = s[-1]
                temp.left = TreeNode(preorder[i])
                s.append(temp.left)
            i += 1
        return root

File: 30-Day LeetCoding Challenge/C++/bstFromPreorder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        int size = preorder.size();
        TreeNode *root = new TreeNode(preorder[0]);
        int i = 1;
        stack<TreeNode*> s;
        s.push(root);
        TreeNode* temp;
        while(i < size) {
            temp = NULL;
            while(!s.empty() && s.top()->val < preorder[i]) {
                temp = s.top();
                s.pop();
            }
            if(temp != NULL) {
                temp->right = new TreeNode(preorder[i]);
                s.push(temp->right);
            }
            else {
                temp = s.top();
                temp->left = new TreeNode(preorder[i]);
                s.push(temp->left);
                
            }
            i++;
        }
        return root;
    }
};